ports/mimxrt/hal/fsl_lpuart.c

```diff
@@ -0,0 +1,2013 @@
+/*
```

Comment:
Which MCU source file is this taken from?  Doesn't it need to be different for each MCU?

Is there a better way to fix the code here rather than copying the file?

Reply:
As far as I recall, that was from the 1062 lib file. But this file is identical for all MIMXRT1xxx MCUs. Newer Versions of the lib use a common UART driver file for all MIMXRT MCU, and this hal/fsl_lpuart.c file is besisdes the changes identical to the one I used a baseline.
I tested it with different MCUs, but can re-test it again.
The change was needed to make IRQ_RXIDLE work as intended by the MCU developers.

Reply:
If you tested it on different MCUs then I'm happy with that.

Reply:
I repeated the test with MIMXRT1010, MIMXRT1015, MIMXRT1021, MIMXRT1052, MIMXRT1062 and MIMXRT1172. All work fine. The test script is set up for Teensy 4.x, but that restriction can be dropped, since the MIMXRT port does not need to specify UART pins.

Reply:
Very good.

Reply:
Hi @robert-hh I'm looking at this file from the perspective of pulling in a new version of it for SDK update. With the auto-formatting done to micropython code it's hard to be sure of the changes you've made.
Was it just the changes in `LPUART_TransferHandleIRQ()` where the large block in the middle `/* If IDLE flag is set and the IDLE interrupt is enabled. */` was removed and replaced with the smaller version at the end?

Before:
``` c
    /* If IDLE flag is set and the IDLE interrupt is enabled. */
    if ((0U != ((uint32_t)kLPUART_IdleLineFlag & status)) &&
        (0U != ((uint32_t)kLPUART_IdleLineInterruptEnable & enabledInterrupts)))
    {
#if defined(FSL_FEATURE_LPUART_HAS_FIFO) && FSL_FEATURE_LPUART_HAS_FIFO
        count = ((uint8_t)((base->WATER & LPUART_WATER_RXCOUNT_MASK) >> LPUART_WATER_RXCOUNT_SHIFT));

        while ((0U != handle->rxDataSize) && (0U != count))
        {
            tempCount = (uint8_t)MIN(handle->rxDataSize, count);

            /* Using non block API to read the data from the registers. */
            LPUART_ReadNonBlocking(base, handle->rxData, tempCount);
            handle->rxData = &handle->rxData[tempCount];
            handle->rxDataSize -= tempCount;
            count -= tempCount;

            /* If rxDataSize is 0, invoke rx idle callback.*/
            if (0U == (handle->rxDataSize))
            {
                handle->rxState = (uint8_t)kLPUART_RxIdle;

                if (NULL != handle->callback)
                {
                    handle->callback(base, handle, kStatus_LPUART_RxIdle, handle->userData);
                }
            }
        }
#endif
        /* Clear IDLE flag.*/
        base->STAT = ((base->STAT & 0x3FE00000U) | LPUART_STAT_IDLE_MASK);

        /* If rxDataSize is 0, disable rx ready, overrun and idle line interrupt.*/
        if (0U == handle->rxDataSize)
        {
            /* Disable and re-enable the global interrupt to protect the interrupt enable register during
             * read-modify-wrte. */
            irqMask = DisableGlobalIRQ();
            base->CTRL &= ~(uint32_t)(LPUART_CTRL_RIE_MASK | LPUART_CTRL_ILIE_MASK | LPUART_CTRL_ORIE_MASK);
            EnableGlobalIRQ(irqMask);
        }
        /* Invoke callback if callback is not NULL and rxDataSize is not 0. */
        else if (NULL != handle->callback)
        {
            handle->callback(base, handle, kStatus_LPUART_IdleLineDetected, handle->userData);
        }
        else
        {
            /* Avoid MISRA 15.7 */
        }
    }
```
after 
``` c
    /* If IDLE flag is set and the IDLE interrupt is enabled. */
    if ((0U != ((uint32_t)kLPUART_IdleLineFlag & status)) &&
        (0U != ((uint32_t)kLPUART_IdleLineInterruptEnable & enabledInterrupts))) {
        /* Clear IDLE flag.*/
        base->STAT |= LPUART_STAT_IDLE_MASK;
        if (NULL != handle->callback) {
            handle->callback(base, handle, kStatus_LPUART_IdleLineDetected, handle->userData);
        } else {
            /* Avoid MISRA 15.7 */
        }
    }
```

Reply:
Comparinf is easier if you run the initial fsl_lpuart.c through uncrustify as well before comparing. So yes, the changes focus on handling the line_idle signal. I removed most of the initial code dealing with it and just kept the part which handles the callback mechanism. So it looks like the code you show is identical to the blocks below.
So this block is removed, which is an `if() .....`:
```
    /* If IDLE flag is set and the IDLE interrupt is enabled. */
    if ((0U != ((uint32_t)kLPUART_IdleLineFlag & status)) &&
        (0U != ((uint32_t)kLPUART_IdleLineInterruptEnable & enabledInterrupts))) {
        #if defined(FSL_FEATURE_LPUART_HAS_FIFO) && FSL_FEATURE_LPUART_HAS_FIFO
        count = ((uint8_t)((base->WATER & LPUART_WATER_RXCOUNT_MASK) >> LPUART_WATER_RXCOUNT_SHIFT));

        while ((0U != handle->rxDataSize) && (0U != count)) {
            tempCount = (uint8_t)MIN(handle->rxDataSize, count);

            /* Using non block API to read the data from the registers. */
            LPUART_ReadNonBlocking(base, handle->rxData, tempCount);
            handle->rxData = &handle->rxData[tempCount];
            handle->rxDataSize -= tempCount;
            count -= tempCount;

            /* If rxDataSize is 0, invoke rx idle callback.*/
            if (0U == (handle->rxDataSize)) {
                handle->rxState = (uint8_t)kLPUART_RxIdle;

                if (NULL != handle->callback) {
                    handle->callback(base, handle, kStatus_LPUART_RxIdle, handle->userData);
                }
            }
        }
        #endif
        /* Clear IDLE flag.*/
        base->STAT = ((base->STAT & 0x3FE00000U) | LPUART_STAT_IDLE_MASK);

        /* If rxDataSize is 0, disable rx ready, overrun and idle line interrupt.*/
        if (0U == handle->rxDataSize) {
            /* Disable and re-enable the global interrupt to protect the interrupt enable register during
             * read-modify-wrte. */
            irqMask = DisableGlobalIRQ();
            base->CTRL &= ~(uint32_t)(LPUART_CTRL_RIE_MASK | LPUART_CTRL_ILIE_MASK | LPUART_CTRL_ORIE_MASK);
            EnableGlobalIRQ(irqMask);
        }
        /* Invoke callback if callback is not NULL and rxDataSize is not 0. */
        else if (NULL != handle->callback) {
            handle->callback(base, handle, kStatus_LPUART_IdleLineDetected, handle->userData);
        } else {
            /* Avoid MISRA 15.7 */
        }
    }
```
And this on is added, which was part of the removed block:
```
    /* If IDLE flag is set and the IDLE interrupt is enabled. */
    if ((0U != ((uint32_t)kLPUART_IdleLineFlag & status)) &&
        (0U != ((uint32_t)kLPUART_IdleLineInterruptEnable & enabledInterrupts))) {
        /* Clear IDLE flag.*/
        base->STAT |= LPUART_STAT_IDLE_MASK;
        if (NULL != handle->callback) {
            handle->callback(base, handle, kStatus_LPUART_IdleLineDetected, handle->userData);
        } else {
            /* Avoid MISRA 15.7 */
        }
    }
```